### PR TITLE
Fix delayed operation unexpected cancellation

### PR DESCRIPTION
**Pull request by aprochwicz**

### CODE REVIEW

File: src/mac_features/nrf_802154_delayed_trx.c [hunk 1]
```diff
@@ -187,19 +197,21 @@ static void nrf_802154_rsch_delayed_rx_timeslot_started(void)
...
             nrf_802154_timer_sched_add(&m_timeout_timer, true);
         }
         else
         {
-            dly_op_stop(RSCH_DLY_RX);
+            dly_op_state_set(RSCH_DLY_RX, DELAYED_TRX_OP_STATE_STOPPED);
         }
     }
-    else
+    if (result)
```
Comment (hubertmis) [inline src/mac_features/nrf_802154_delayed_trx.c]: Did you mean `if (!result)` ?

File: src/mac_features/nrf_802154_delayed_trx.c [hunk 2]
```diff
@@ -52,50 +52,53 @@
...
+ * @brief States of delayed operations.
  */
-static void dly_op_start(rsch_dly_ts_id_t dly_ts_id)
+typedef enum
 {
-    assert(dly_ts_id < RSCH_DLY_TS_NUM);
+    DELAYED_TRX_OP_STATE_STOPPED, ///< Delayed operation stopped.
+    DELAYED_TRX_OP_STATE_PENDING, ///< Delayed operation scheduled and waiting for timeslot.
+    DELAYED_TRX_OP_STATE_ONGOING, ///< Delayed operation ongoing (during timeslot).
+    DELAYED_TRX_OP_STATE_NB       ///< Number of delayed operation states.
```
[the remark this answers is not in the document]
Comment (aprochwicz) [reply src/mac_features/nrf_802154_delayed_trx.c]: Supposed to by `NBR`, but it seems to be very rarely used abbreviation.

[the remark this answers is not in the document]
Comment (hubertmis) [reply src/mac_features/nrf_802154_delayed_trx.c]: North British Railway? ;) I'm ok with this, just never seen in software before.

File: src/mac_features/nrf_802154_delayed_trx.c [hunk 3]
```diff
@@ -187,19 +197,21 @@ static void nrf_802154_rsch_delayed_rx_timeslot_started(void)
                                             true);
         if (result)
         {
+            dly_op_state_set(RSCH_DLY_RX, DELAYED_TRX_OP_STATE_ONGOING);
```
[the remark this answers is not in the document]
Comment (aprochwicz) [reply src/mac_features/nrf_802154_delayed_trx.c]: Moved state and timer handling to receive request status callback.

File: src/mac_features/nrf_802154_delayed_trx.c [hunk 4]
```diff
@@ -49,79 +49,143 @@
...
 /**
- * Start delayed timeslot operation.
+ * Set state of a delayed operation.
  *
- * @param[in]  dly_ts_id  Delayed timeslot ID.
+ * @param[in]  dly_ts_id    Delayed timeslot ID.
+ * @param[in]  dly_op_state Delayed operation state.
  */
-static void dly_op_start(rsch_dly_ts_id_t dly_ts_id)
+void dly_op_state_set(rsch_dly_ts_id_t dly_ts_id, delayed_trx_op_state_t dly_op_state)
```
Comment (hubertmis) [inline src/mac_features/nrf_802154_delayed_trx.c]: `static` is missing

File: src/mac_features/nrf_802154_delayed_trx.c [hunk 5]
```diff
@@ -49,79 +49,143 @@
...
 /**
- * Stop delayed timeslot operation.
+ * Get state of a delayed operation.
  *
  * @param[in]  dly_ts_id  Delayed timeslot ID.
+ *
+ * @retval     State of delayed operation.
  */
-static void dly_op_stop(rsch_dly_ts_id_t dly_ts_id)
+delayed_trx_op_state_t dly_op_state_get(rsch_dly_ts_id_t dly_ts_id)
```
Comment (hubertmis) [inline src/mac_features/nrf_802154_delayed_trx.c]: `static` is missing

File: src/mac_features/nrf_802154_delayed_trx.c [hunk 6]
```diff
@@ -209,13 +296,11 @@ bool nrf_802154_delayed_trx_transmit(const uint8_t * p_data,
                                      uint32_t        dt,
                                      uint8_t         channel)
 {
-    bool     result = true;
+    bool     result;
     uint16_t timeslot_length;
+    bool     acq;
```
Comment (hubertmis) [inline src/mac_features/nrf_802154_delayed_trx.c]: What `acq` stands for? Why not `ack` or `ack_req`?

File: src/mac_features/nrf_802154_delayed_trx.c [hunk 7]
```diff
@@ -257,102 +330,94 @@ bool nrf_802154_delayed_trx_receive(uint32_t t0,
...
-    else if ((REQ_ORIG_HIGHER_LAYER == req_orig) || (term_lvl >= NRF_802154_TERM_802154))
-    {
-        nrf_802154_timer_sched_remove(&m_timeout_timer);
-        dly_op_stop(RSCH_DLY_RX);
-    }
-    else
+    if (dly_op_state_get(RSCH_DLY_RX) == DELAYED_TRX_OP_STATE_ONGOING)
     {
-        result = false;
+        if ((REQ_ORIG_HIGHER_LAYER == req_orig) || (term_lvl >= NRF_802154_TERM_802154))
```
Comment (hubertmis) [inline src/mac_features/nrf_802154_delayed_trx.c]: Is it needed to terminate delayed rx operation whenever higher layer requests it? Regardless `term_lvl` value?

File: src/mac_features/nrf_802154_delayed_trx.c (hunk 6 above)
Comment (aprochwicz) [reply src/mac_features/nrf_802154_delayed_trx.c]: Changed to `ack`.

File: src/mac_features/nrf_802154_delayed_trx.c (hunk 4 above)
Comment (aprochwicz) [reply src/mac_features/nrf_802154_delayed_trx.c]: Added static.

File: src/mac_features/nrf_802154_delayed_trx.c (hunk 7 above)
Comment (aprochwicz) [reply src/mac_features/nrf_802154_delayed_trx.c]: Changed to check only `term_lvl`.

**Comment by aprochwicz** [comment]
Regression test

**Comment by aprochwicz** [comment]
Regression test

**Pull request merged**